docs/DynamicCasting.md

```diff
@@ -0,0 +1,532 @@
+# Dynamic Casting Behavior
+
+* Author: [Tim Kientzle](https://github.com/tbkka)
+* Implementation: [apple/swift#29658](https://github.com/apple/swift/pull/29658)
+
+## Introduction
+
+The Swift language has three casting operators: `is`, `as?`, and `as!`.
+Each one takes an instance on the left-hand side and a type expression on the right-hand side.
+
+* The _cast test operator_ `is` tests whether a particular instance can be converted to a particular destination type.  It returns a boolean result.
+
+* The _conditional cast operator_ `as?` attempts the conversion and returns an `Optional` result: `nil` if the conversion failed, and a non-nil optional containing the result otherwise.
+
+* The _forced cast operator_ `as!` unconditionally performs the casting conversion and returns the result.  If an `as!` expression does not succeed, the implementation may terminate the program.
+
+Note: The static coercion operator `as` serves a different role and it's behavior will not be specifically discussed in this document.
+
+The following invariants relate the three casting operators:
+* Cast test: `x is T == ((x as? T) != nil)`
+* Conditional cast: `(x as? T) == (x is T) ? .some(x as! T) : nil`
+* Forced cast: `x as! T` is equivalent to `(x as? T)!`
+
+In particular, note that `is` and `as!` can be implemented in terms of `as?` and vice-versa.
+
+As with other operators with `!` in their names, `as!` is intended to only be invoked in cases where the programmer knows a priori that the cast will succeed.
+If the conversion would not succeed, then the behavior may not be fully deterministic.
+See the discussion of Array casting below for a specific case where this non-determinism can be important.
+
+The following sections detail the rules that govern the casting operations for particular Swift types.
+Except as described below, casting between dissimilar types always fails -- for example, casting a struct instance to a function type or vice-versa.
+
+Where possible, each section includes machine-verifiable _invariants_ that can be used as the basis for developing a robust test suite for this functionality.
+
+## Identity Cast
+
+Casting an instance of a type to its own type will always succeed and return the original value unchanged.
+
+```
+let a: Int = 7
+a is Int // true
+a as? Int // Succeeds
+a as! Int == a // true
```

Comment:
Swift has a nice set of warnings for each of these casts that could be highlighted.

Reply:
I fear that would actually detract from the point here.
Formally, I need these as the base case for later invariants of the form
    `(a .....) is Int == a is Int`
Quoting compiler warnings could give someone the impression that identity casts are somehow broken or unsound, rather than simply unnecessary.